src/App.tsx

```diff
@@ -36,35 +50,63 @@ function App(props: Props) {
       state.signer &&
       state.address &&
       roundId &&
-      (previousAddress === null || previousSigner === null)
+      (!previousAddress || !previousSigner)
     ) {
       const address = state.address;
       const message = `UMA Protocol one time key for round: ${roundId}`;
+      const hexMessage = web3.utils.toHex(message);
       const keyExists = signingKeys[address];
-      if (!keyExists || keyExists.roundMessage !== message) {
-        state.signer
-          .signMessage(message)
-          .then((msg) => {
-            const key = {} as {
-              publicKey: string;
-              privateKey: string;
-              roundMessage: string;
-            };
-
-            const privateKey = derivePrivateKey(msg);
-            const publicKey = recoverPublicKey(privateKey);
-            key.privateKey = privateKey;
-            key.publicKey = publicKey;
-            key.roundMessage = message;
-
-            setSigningKeys((prevKeys) => {
-              return { ...prevKeys, [address]: key };
+      let keysInStorage =
+        localStorage.getItem(SIGNING_KEYS_STORAGE_KEY) || ({} as StorageKeys);
+
+      let keyExistsInStorage = false;
+      if (typeof keysInStorage === "string") {
+        keysInStorage = JSON.parse(keysInStorage) as StorageKeys;
+
+        keyExistsInStorage = has(keysInStorage, `${hexMessage}.${address}`);
+      }
+
+      if (keyExistsInStorage) {
+        const loggedInKey = keysInStorage[hexMessage][address];
+        setSigningKeys((prevKeys) => {
+          return { ...prevKeys, [address]: loggedInKey };
+        });
+      } else {
+        if (!keyExists || keyExists.roundMessage !== message) {
+          state.signer
+            .signMessage(message)
+            .then((msg) => {
+              const key = {} as SigningKey;
+
+              const privateKey = derivePrivateKey(msg);
+              const publicKey = recoverPublicKey(privateKey);
+              key.privateKey = privateKey;
+              key.publicKey = publicKey;
+              key.roundMessage = message;
+              key.roundId = roundId;
+
+              let keysToBackup = {};
+              if (typeof keysInStorage === "string") {
+                const parsedJSON = JSON.parse(keysInStorage);
+                keysToBackup = { ...parsedJSON };
+              } else {
+                keysToBackup = { ...keysInStorage };
+              }
+
+              setWith(keysToBackup, `${hexMessage}.${address}`, key, Object);
+              localStorage.setItem(
+                SIGNING_KEYS_STORAGE_KEY,
+                JSON.stringify(keysToBackup)
+              );
+              setSigningKeys((prevKeys) => {
+                return { ...prevKeys, [address]: key };
+              });
+            })
+            .catch((err) => {
+              const error = new Error("Sign failed.");
```

Comment:
this obscures the actual error, what do u think about doing something like this `new Error("Sign failed: " + err.message);`

Reply:
Yep, we can do that. This was largely initially for when a user clicked cancel, but we can just return the error message.

Reply:
it looks like this would mainly throw in the signer, but it could also throw in business logic for whatever reason.  we could also just log to console.error if you dont want to expose the user to the raw message.  just need a way to debug if someoen reports a problem imo